README.md
# Visibility.js: a wrapper for the Page Visibility API

Visibility.js allows you to determine whether your web page is visible to a
user, is hidden in background tab or is prerendering. It allows you to use
the page visibility state in JavaScript logic and improve browser performance
by disabling unnecessary timers and AJAX requests, or improve user interface
experience (for example, by stopping video playback or slideshow when user
switches to another browser tab).

Moreover, you can detect if the browser is just [prerendering] the page while
the user has not still opened the link, and don’t count this as a visit in your
analytics module, or do not run heavy calculations or other actions which will
disable the prerendering.

This library is a wrapper of the [Page Visibility API]. It eases usage
of the API by hiding vendor-specific property prefixes and adding some
high-level functions.

In most cases you don’t need to check whether the Page Visibility API is
actually supported in the browser as, if it does not, the library will just
assume that the page is visible all the time, and your logic will still work
correctly, albeit less effective in some cases.

Page Visibility API is [natively supported] by all browsers. For old browsers
you can use `lib/visibility.fallback.js` with focus/blur hack (note that this
hack have issue, that document become to be hidden, when browser just
lose focus, but still visible for user).

Sponsored by [Evil Martians].

[Page Visibility API]: http://www.w3.org/TR/page-visibility/
[prerendering]:        http://code.google.com/chrome/whitepapers/prerender.html
[natively supported]:  http://caniuse.com/pagevisibility
[Evil Martians]:       http://evilmartians.com/

## Translations

Документация на русском:
[habrahabr.ru/blogs/javascript/125833/](http://habrahabr.ru/blogs/javascript/125833/)

## States

Currently the Page Visibility API supports three visibility states:

* `visible`: user has opened the page and works within it.
* `hidden`: user has switched to another tab or minimized browser window.
* `prerender`: browser is just prerendering a page which may possibly be opened
   by the user to make the apparent loading time lesser.

## Timers

The main use case for this library is to enable some of the times only when
content is visible to the user, i.e. the ones animating a countdown animation.

`Visibility.every(interval, callback)` is similar to
`setInterval(callback, interval)`, but calls `callback` every `interval` ms only
if the page is visible. For example, let’s create a countdown timer:

```js
Visibility.every(1000, function () {
    updateCountdownAnimation();
});
```

You can provide an additional interval which will be used when the page
is hidden. In next example, a check for inbox updates will be run every 1 minute
for a visible page and every 5 minutes for a hidden one:

```js
var minute = 60 * 1000;
Visibility.every(minute, 5 * minute, function () {
    checkForEmail();
});
```

Note that the callback will also be executed on every `hidden`->`visible` state
change to update old contents.

`Visibility.every` returns a timer identifier, much like the `setTimeout`
function. It cannot be passed to `clearInterval`, through, and you should use
`Visibility.stop(id)` to stop the timer.

```js
var slideshow = Visibility.every(5 * 1000, function () {
    nextSlide();
});

$('.stopSlideshow').click(function () {
    Visibility.stop(slideshow);
});
```

If the browser does not support the Page Visibility API, `Visibility.every` will
fall back to `setInterval`, and `callback` will be run every `interval` ms for
both the hidden and visible pages.

## Initializers

In another common use case you need to execute some actions upon a switch to
particular visibility state.

### Waiting until the page becomes visible

`Visibility.onVisible(callback)` checks current state of the page. If it is
visible now, it will run `callback`, otherwise it will wait until state changes
to `visible`, and then run `callback`.

For example, let’s show an animated notification only when the page is visible,
so if an user opens a page in the background, the animation will delay until
the page becomes visible, i.e. until the user has switched
to a tab with the page:

```js
Visibility.onVisible(function () {
    startIntroAnimation();
});
```

If a browser doesn’t support Page Visibility API, `Visibility.onVisible`
will run the `callback` immediately.

### Wait until the page is opened after prerendering

A web developer can hint a browser (using Prerendering API) that an user
is likely to click on some link (i.e. on a “Next” link in a multi-page article),
and the browser then may prefetch and prerender the page, so that the user will
not wait after actually going via the like.

But you may not want to count the browser prerendering a page as a visitor in
your analytics system. Moreover, the browser will disable prerendering if you
will try to do heavy computations or use audio/video tags on the page. So, you
may decide to not run parts of the code while prerendering and wait until the
user actually opens the link.

You can use `Visibility.afterPrerendering(callback)` in this cases. For example,
this code will only take real visitors (and not page prerenderings) into
account:

```js
Visibility.afterPrerendering(function () {
    Statistics.countVisitor();
});
```

If the browser doesn’t support Page Visibility API,
`Visibility.afterPrerendering` will run `callback` immediately.

## Low-level API

In some cases you may need more low-level methods. For example, you may want to
count the time user has viewed the page in foreground and time it has stayed in
background.

`Visibility.isSupported()` will return `true` if browser supports the
Page Visibility API:

```js
if( Visibility.isSupported() ) {
    Statistics.startTrackingVisibility();
}
```

`Visibility.state()` will return a string with visibility state. More states
can be added in the future, so for most cases a simpler `Visibility.hidden()`
method can be used. It will return `true` if the page is hidden by any reason.
For example, while prerendering, `Visibility.state()` will return `"prerender"`,
but `Visibility.hidden()` will return `true`.

This code will aid in collecting page visibility statistics:

```js
$(document).load(function () {

    if ( 'hidden' == Visibility.state() ) {
        Statistics.userOpenPageInBackgroundTab();
    }
    if ( 'prerender' == Visibility.state() ) {
        Statistics.pageIsPrerendering();
    }

});
```

And this example will only enable auto-playing when the page is opening as a
visible tab (not a background one):

```js
$(document).load(function () {

   if ( !Visibility.hidden() ) {
       VideoPlayer.play();
   }

});
```

Using `Visibility.change(callback)` you can listen to visibility state changing
events. The `callback` takes 2 arguments: an event object and a state name.

Let’s collect some statistics with this evented approach:

```js
Visibility.change(function (e, state) {
    Statistics.visibilityChange(state);
});
```

Method `change` returns listener ID. You can use it to unbind listener by
`Visibility.unbind(id)`:

```js
var listener = Visibility.change(function (e, state) {
    if ( !Visibility.hidden() ) {
       VideoPlayer.pause();
    }
});

VideoPlayer.onFinish(function () {
    Visibility.unbind(listener);
});
```

Methods `onVisible` and `afterPrerendering` will also return listener ID,
if they wait visibility state changes. If they execute callback immediately,
they return `true` if Page Visibility API is supported and `false`
if they can’t detect visibility state.

```js
var listener = Visibility.onVisible(function () {
    notification.takeAttention();
});

notification.onOutOfDate(function () {
    if ( typeof(listener) == 'number' ) {
        Visibility.unbind(listener);
    }
});
```

## Installing

### Packages

Visibility.js is shipped with 4 files:

* `visibility.core` – core module.
* `visibility.timers` – `every` and `stop` methods to set `setTimeout` depend
  on visibility state.
* `visibility` – `visibility.core` and `visibility.timers` together.
* `visibility.fallback` – fallback for browser without Page Visibility API.
  It use document `focus`/`blur` events, so document become to be hidden,
  when browser just lose focus, but still visible for user.

### Ruby on Rails

For Ruby on Rails you can use gem for Assets Pipeline.

1. Add `visibilityjs` gem to `Gemfile`:

   ```ruby
   gem "visibilityjs"
   ```

2. Install gems:

   ```sh
   bundle install
   ```

3. Include Visibility.js in your `application.js.coffee`:

   ```coffee
   #= require visibility
   ```

   If you willn’t use `every` method, you can reduce library size by including
   only core module:

   ```coffee
   #= require visibility.core
   ```

### CDN

If you don’t use any assets packaging manager use [cdnjs](http://cdnjs.com/).
Add to your site:

```html
<script src="//cdnjs.cloudflare.com/ajax/libs/visibility.js/1.0.0/visibility.min.js"></script>
```

### Other

If you need just a files, you can take already minified packages from
[github.com/ai/visibility.js/releases](https://github.com/ai/visibility.js/releases).

## Contributing

1. To run tests you need node.js and npm. For example, in Ubuntu run:

   ```sh
   sudo apt-get install nodejs npm
   ```

2. Next install npm dependencies:

   ```sh
   npm install
   ```

3. Run all tests:

   ```sh
   npm test
   ```

4. Run test server, to check code in real browsers:

   ```sh
   ./node_modules/.bin/cake server
   ```

5. Open tests in browser: [localhost:8000](http://localhost:8000).
6. Also you can see real usage example in integration test
   `test/integration.html`.


ChangeLog
== 1.0.0 (Discoverer 2, stable)
* Remove jQuery.Chrono integration.
* Remove outdated Firefox prefix.

== 0.6.2 (Pioneer 4, american sun)
* Decrease files size (by compressible code and UnglifyJS 2).
* Remove unnecessary vendor prefixes from fallback.

== 0.6.1 (Vanguard 2, weather)
* Remove unnecessary vendor prefixes.

== 0.6 (Luna 1, Mechta)
* Methods onVisible and afterPrerendering return listener ID (by mcfedr).
* Fix documentation (by Erwann Mest).

== 0.5 (SCORE, communication)
* Split library to core and timers modules.
* Allow to unbind change listener.
* Use common logic in change(), afterPrerendering() and onVisible().

== 0.4.5 (Pioneer 3, closer)
* Fix gem assets directory.
* Update development dependencies.

== 0.4.4 (Pioneer 1, deeper into space)
* Remove non-ASCII symbols from gemspec.
* Print testing URL in test task.

== 0.4.3 (Explorer 4, short)
* Fix gemspec issue with Bundler.

== 0.4.2 (Sputnik 3, real)
* Reorder code to show first public and common methods.
* Use node.js Cake instead of Ruby’s Rake to build tasks.
* Move to Mocha, Chai and Sinon.JS for tests.
* Move autogenerated minified to GitHub downloads.

== 0.4.1 (Explorer 3, repeat)
* Fix documentation and gemspec.

== 0.4 (Vanguard 1, alternative)
* Add fallback API support by focus/blur hack to all browsers.

== 0.3 (Explorer 1, answer)
* Rename gem to visibilityjs.
* Fix gem integration with Asset Pipeline paths.

== 0.2 (Sputnik 2, Laika)
* Rename support() to isSupported() and notPrerender() to afterPrerendering().
* Fix IE 6 and 7 support.
* Documentation fixes by Peter Zotov.

== 0.1 (Sputnik 1, the first)
* Initial release.


lib/visibility.js
/*
 * Copyright 2011 Andrey “A.I.” Sitnik <andrey@sitnik.ru>,
 * sponsored by Evil Martians.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

;(function() {
    "use strict";

    var defined = function(variable) {
        return ('undefined' != typeof(variable));
    };

    // Visibility.js allow you to know, that your web page is in the background
    // tab and thus not visible to the user. This library is wrap under
    // Page Visibility API. It fix problems with different vendor prefixes and
    // add high-level useful functions.
    window.Visibility = {

        // Call callback only when page become to visible for user or
        // call it now if page is visible now or Page Visibility API
        // doesn’t supported.
        //
        // Return true if callback if called now.
        //
        //   Visibility.onVisible(function() {
        //       Notification.animateNotice("Hello");
        //   });
        onVisible: function (callback) {
            if ( !this.isSupported() || !this.hidden() ) {
                callback();
                return true;
            }
            this._onVisibleCallbacks.push(callback);
            this._setListener();
        },

        // Run callback every `interval` milliseconds if page is visible and
        // every `hiddenInterval` milliseconds if page is hidden.
        //
        //   Visibility.every(60 * 1000, 5 * 60 * 1000, function() {
        //       checkNewMails();
        //   });
        //
        // You can skip `hiddenInterval` and callback will be called only if
        // page is visible.
        //
        //   Visibility.every(1000, function() {
        //       updateCountdown();
        //   });
        //
        // It is analog of `setInterval(callback, interval)` but use visibility
        // state.
        //
        // It return timer ID, that you can use in `Visibility.stop(id)` to stop
        // timer (`clearInterval` analog).
        // Warning: timer ID is different from intervalID from `setInterval`,
        // so don’t use it in `clearInterval`.
        //
        // On change state from hidden to visible timers will be execute.
        //
        // If you include jQuery Chrono plugin before Visibility.js, you could
        // use Chrono’s syntax sugar in interval arguments:
        //
        //   Visibility.every('second', function() {
        //       updateCountdown();
        //   });
        //   Visibility.every('1 minute', '5 minutes', function() {
        //       checkNewMails();
        //   });
        every: function (interval, hiddenInterval, callback) {
            if ( !defined(callback) ) {
                callback = hiddenInterval;
                hiddenInterval = null;
            }
            this._lastTimer += 1;
            var number = this._lastTimer;
            this._timers[number] = ({
                interval:       interval,
                hiddenInterval: hiddenInterval,
                callback:       callback
            });
            this._runTimer(number, false);
            if ( this.isSupported() ) {
                this._setListener();
            }
            return number;
        },

        // Stop timer from `every` method by it ID (`every` method return it).
        //
        //   slideshow = Visibility.every(5 * 1000, function() {
        //       changeSlide();
        //   });
        //   $('.stopSlideshow').click(function() {
        //       Visibility.stop(slideshow);
        //   });
        stop: function(id) {
            var timer = this._timers[id]
            if ( !defined(timer) ) {
                return false;
            }
            this._stopTimer(id);
            delete this._timers[id];
            return timer;
        },

        // Call callback when visibility will be changed. First argument for
        // callback will be original event object, second will be visibility
        // state name.
        //
        // If Page Visibility API doesn’t supported method will be return false
        // and callback never will be called.
        //
        //   Visibility.change(function(e, state) {
        //       Statistics.visibilityChange(state);
        //   });
        //
        // It is just proxy to `visibilitychange` event, but use vendor prefix.
        change: function (callback) {
            if ( !this.isSupported() ) {
                return false;
            }
            this._changeCallbacks.push(callback);
            this._setListener();
            return true;
        },

        // Call `callback` in any state, expect “prerender”. If current state
        // is “prerender” it will wait until state will be changed.
        // If Page Visibility API doesn’t supported, it will call `callback`
        // immediately.
        //
        //   Visibility.afterPrerendering(function () {
        //       Statistics.countVisitor();
        //   });
        afterPrerendering: function (callback) {
            if ( !this.isSupported() || 'prerender' != this.state() ) {
                callback();
                return true;
            }
            this._afterPrerenderingCallbacks.push(callback);
            this._setListener();
        },

        // Return true if page now isn’t visible to user.
        //
        //   if ( !Visibility.hidden() ) {
        //       VideoPlayer.play();
        //   }
        //
        // It is just proxy to `document.hidden`, but use vendor prefix.
        hidden: function () {
            if ( !this.isSupported() ) {
                return false;
            }
            return this._prop('hidden');
        },

        // Return visibility state: 'visible', 'hidden' or 'prerender'.
        //
        //   if ( 'prerender' == Visibility.state() ) {
        //       Statistics.pageIsPrerendering();
        //   }
        //
        // Don’t use `Visibility.state()` to detect, is page visible, because
        // visibility states can extend in next API versions.
        // Use more simpler and general `Visibility.hidden()` for this cases.
        //
        // It is just proxy to `document.visibilityState`, but use
        // vendor prefix.
        state: function () {
            if ( !this.isSupported() ) {
                return 'visible';
            }
            return this._prop('visibilityState');
        },

        // Return true if browser support Page Visibility API.
        //
        //   if ( Visibility.isSupported() ) {
        //       Statistics.startTrackingVisibility();
        //       Visibility.change(function(e, state)) {
        //           Statistics.trackVisibility(state);
        //       });
        //   }
        isSupported: function () {
            return defined(this._prefix());
        },

        // Link to document object to change it in tests.
        _doc: window.document,

        // Vendor prefixes to create event and properties names.
        _prefixes: ['webkit', 'moz', 'o', 'ms'],

        // Vendor prefix cached by `_prefix` function.
        _chechedPrefix: null,

        // Is listener for `visibilitychange` event is already added
        // by `_setListener` method.
        _listening: false,

        // Callbacks from `change` method, that wait visibility changes.
        _changeCallbacks: [],

        // Callbacks from `onVisible` method, that wait when page become to be
        // visible.
        _onVisibleCallbacks: [],

        // Callbacks from `afterPrerendering` method, that wait when visibility
        // state change from “prerender”.
        _afterPrerenderingCallbacks: [],

        // Last timer number.
        _lastTimer: 0,

        // Callbacks and intervals added by `every` method.
        _timers: { },

        // Variable to check hidden-visible state changes.
        _hiddenBefore: false,

        // Initialize variables on page loading.
        _init: function () {
            this._hiddenBefore = this.hidden();

            if ( defined(window.jQuery) && defined(jQuery.every) ) {
                this._setInterval = this._chronoSetInterval;
            } else {
                this._setInterval = this._originalSetInterval;
            }
        },

        // Detect vendor prefix and return it.
        _prefix: function () {
            if ( null !== this._chechedPrefix ) {
                return this._chechedPrefix;
            }
            if ( defined(this._doc.visibilityState) ) {
                return this._chechedPrefix = '';
            }
            var name;
            for ( var i = 0; i < this._prefixes.length; i++ ) {
                name = this._prefixes[i] + 'VisibilityState';
                if ( defined(this._doc[name]) ) {
                    return this._chechedPrefix = this._prefixes[i];
                }
            }
        },

        // Return property name with vendor prefix.
        _name: function (name) {
            var prefix = this._prefix();
            if ( '' == prefix ) {
                return name;
            } else {
                return prefix +
                    name.substr(0, 1).toUpperCase() + name.substr(1);
            }
        },

        // Return document’s property value with name with vendor prefix.
        _prop: function (name) {
            return this._doc[this._name(name)]
        },

        // Listener for `visibilitychange` event.
        _onVisibilityChange: function(event) {
            var isHidden = this.hidden(),
                state    = this.state();

            for ( var i = 0; i < this._changeCallbacks.length; i++ ) {
                this._changeCallbacks[i].call(this._doc, event, state);
            }

            var hiddenBefore = this._hiddenBefore;
            if ( (isHidden && !hiddenBefore) || (!isHidden && hiddenBefore) ) {
                for ( i in this._timers ) {
                    this._stopTimer(i);
                    this._runTimer(i, !isHidden);
                }
            }

            if ( !isHidden ) {
                for ( var i = 0; i < this._onVisibleCallbacks.length; i++ ) {
                    this._onVisibleCallbacks[i]();
                }
                this._onVisibleCallbacks = [];
            }

            if ( 'prerender' != this.state() ) {
                var after = this._afterPrerenderingCallbacks.length
                for ( var i = 0; i < after; i++ ) {
                    this._afterPrerenderingCallbacks[i]();
                }
                this._afterPrerenderingCallbacks = [];
            }

            this._hiddenBefore = isHidden;
        },

        // Set listener for `visibilitychange` event.
        _setListener: function () {
            if ( this._listening ) {
                return;
            }
            var event = this._prefix() + 'visibilitychange';
            var listener = function () {
                Visibility._onVisibilityChange.apply(Visibility, arguments);
            };
            if ( this._doc.addEventListener ) {
                this._doc.addEventListener(event, listener, false);
            } else {
                this._doc.attachEvent(event, listener);
            }
            this._listening = true;
            this._hiddenBefore = this.hidden();
        },

        // Set interval directly by `setInterval` function without any syntax
        // sugar.
        _originalSetInterval: function (callback, interval) {
            return setInterval(callback, interval);
        },

        // Set interval by jQuery Chrono plugin. Add syntax sugar to `interval`
        // and `hiddenInterval` arguments, such as "1 second" and others.
        //
        // It will be automatically set to `_setInterval` on loading if
        // you include jQuery Chrono plugin before Visibility.js.
        _chronoSetInterval: function (callback, internal) {
            return jQuery.every(internal, callback);
        },

        // Set interval by `setInterval`. Allow to change function for tests or
        // syntax sugar in `interval` arguments.
        //
        // Function will be automatically set in `_init` method (which will be
        // call on script loading). So you must include jQuery Chrono plugin
        // before Visibility.js.
        _setInterval: null,

        // Try to run timer from every method by it’s ID. It will be use
        // `interval` or `hiddenInterval` depending on visibility state.
        // If page is hidden and `hiddenInterval` is null,
        // it will not run timer.
        //
        // Argument `now` say, that timers must be execute now too.
        _runTimer: function (id, now) {
            var interval,
                timer = this._timers[id];
            if ( this.hidden() ) {
                if ( null === timer.hiddenInterval ) {
                    return;
                }
                interval = timer.hiddenInterval;
            } else {
                interval = timer.interval;
            }
            if ( now ) {
                timer.callback.call(window);
            }
            timer.intervalID = this._setInterval(timer.callback, interval);
        },

        // Stop timer from `every` method by it’s ID.
        _stopTimer: function (id) {
            var timer = this._timers[id];
            clearInterval(timer.intervalID);
            delete timer.intervalID;
        }

    };

    Visibility._init();

})();
